src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepExecution.java

```diff
@@ -432,7 +432,7 @@ public String getCookie() {
             if (FlowExecutionList.get().isResumptionComplete()) {
                 // We only do this if resumption is complete so that we do not load the run and resume its execution in this context in normal scenarios.
                 Run<?, ?> run = runForDisplay();
-                if (!stopping && run != null && !run.isLogUpdated()) {
+                if (run != null && !run.isLogUpdated()) {
```

Comment:
IDK about this. It was only ever intended to run once after a Jenkins restart (and Damien says that in the current case ci.jenkins.io has not yet restarted), so I am worried that removing the guard may lead to log spam or further confusion. Maybe instead we could add a call to `Timer.execute` [here](https://github.com/jenkinsci/workflow-durable-task-step-plugin/blob/f256ed346587bc61c0cf3aac38f458d768da8a43/src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepExecution.java#L148) that attempts to cancel the item (or maybe its task, if we think that would make a difference) again after a delay.

I think it would be harmless to test this new code, but if ci.jenkins.io really has not yet restarted, I would be interested to see if the previous code is enough to cancel the task after a restart or not.

Reply:
> I think it would be harmless to test this new code, but if ci.jenkins.io really has not yet restarted, I would be interested to see if the previous code is enough to cancel the task after a restart or not.

Oh I'm sorry, our comments crossed: https://github.com/jenkinsci/workflow-durable-task-step-plugin/pull/259#issuecomment-1208240715

Reply:
> I am worried that removing the guard may lead to log spam

I was also concerned about this. But @dduportal said that the script (which mimics what this PR would do) indeed canceled the bogus items successfully.

Reply:
Even if it is fine in this case, it seems clearer and safer to me to trigger an explicit delayed cancellation when the original cancellation attempt fails rather than have `getCauseOfBlockage` potentially cancel things more than once. For example, if the `Timer` thread pool is somewhat congested and the task does not run for more than 5 seconds, I think a second task could be scheduled with the proposed code (not really a big deal).

Either way, at this point it seems that the issue seen on ci.jenkins.io happens while the controller is running rather than being a result of some problem with persisted state after a restart, which I did not know was possible and is distinct from what I was trying to fix when I filed https://github.com/jenkinsci/workflow-durable-task-step-plugin/pull/185. Ideally we would try to reproduce the new issue in a test to have a chance at fixing the underlying cause.

Reply:
> the issue seen on ci.jenkins.io happens while the controller is running rather than being a result of some problem with persisted state after a restart

Yes, that was also my understanding.

> trigger an explicit delayed cancellation when the original cancellation attempt fails

My reservation is that we do not understand the cause of the original failure to cancel, so we cannot know whether that would actually work. The idea with this PR is to be a final fallback in case other things failed for unknown reasons, to make sure the queue items do not remain there indefinitely.

> try to reproduce the new issue in a test

Yes of course that would be best, but I have no clear notion of what happened. `BlockedItem.leave` not working? One case appeared to relate to a Dependabot PR being closed, but that does nothing to explain why `Queue.cancel` would fail.

Not sure what to do at this point other than see how things behave on ci.jenkins.io.

Reply:
I haven't been paying attention too closely but I think I've noticed this on ci.jenkins.io when a PR is closed due to a new build starting, i.e.:

1. A PR build starts
2. A new commit is made
3. A new PR build starts
4. Due to `disableConcurrentBuilds(abortPrevious: true)` the first PR build is aborted
5. The zombie queue items from the first build remain

Reply:
Right, that would be a common trigger for https://github.com/jenkinsci/workflow-durable-task-step-plugin/blob/a62ccc91ca0b593d557038abfc48378939f858bd/src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepExecution.java#L144-L149 to be run but the question is why `cancel` would be returning false (as logs confirm) and the item left in the queue.

Reply:
> My reservation is that we do not understand the cause of the original failure to cancel, so we cannot know whether that would actually work. The idea with this PR is to be a final fallback in case other things failed for unknown reasons, to make sure the queue items do not remain there indefinitely.

Well, we do not really know whether this PR works yet either. As far as we know, the existing code would have fixed the issue when Damien restarted ci.jenkins.io, and we have no way to tell whether the new case is being hit. I think we should at least use a distinct log message for the case where `stopping` is already `true` so that we would know for sure when it applies and since "Refusing to build ..." does not really make sense as an explanation in this scenario anyway (more like "Desperately trying to cancel ...").

Maybe we should use both approaches, at least temporarily, with a distinct log message in each case, just to see what happens. Knowing whether an asynchronous cancellation right after the synchronous cancellation in `stop` works or not could be a useful clue to anyone who ends up trying to track down the root cause in the future.

Reply:
> we do not really know whether this PR works yet either

Not directly, but analogous calls made from `/script` _did_ work, so we have reason to believe it would have.

> As far as we know, the existing code would have fixed the issue when Damien restarted ci.jenkins.io

Yes it probably would have, but we do not want to require admins to restart the server just to clean up items we _know_ are invalid.

> use both approaches, at least temporarily, with a distinct log message

Could try that. @dduportal do you believe this issue happens enough that you would be likely to see it again?

Reply:
https://github.com/jenkins-infra/helpdesk/issues/3211#issuecomment-1317370352